### PR TITLE
Add taint tracking for bufio and bytes packages

### PR DESCRIPTION
**Pull request by gagliardetto**
Part of #167

### CODE REVIEW

**Comment by gagliardetto** [comment]
All test fail here:

```
Executing tests in /home/runner/work/codeql-go/codeql-go/ql/test/library-tests/semmle/go/frameworks/TaintSteps.
--- expected
+++ actual
@@ -37,10 +37,12 @@
 | io.go:79:3:79:8 | reader | io.go:78:3:78:5 | definition of buf |
 | io.go:84:24:84:29 | reader | io.go:84:9:84:33 | call to LimitReader |
 | io.go:85:22:85:23 | lr | io.go:85:11:85:19 | selection of Stdout |
+| io.go:92:3:92:3 | definition of r | io.go:90:3:90:4 | definition of r2 |
 | io.go:92:23:92:24 | r1 | io.go:92:8:92:33 | call to MultiReader |
 | io.go:92:27:92:28 | r2 | io.go:92:8:92:33 | call to MultiReader |
 | io.go:92:31:92:32 | r3 | io.go:92:8:92:33 | call to MultiReader |
 | io.go:93:22:93:22 | r | io.go:93:11:93:19 | selection of Stdout |
+| io.go:98:3:98:5 | definition of tee | io.go:98:26:98:29 | &... |
 | io.go:98:23:98:23 | r | io.go:98:10:98:30 | call to TeeReader |
 | io.go:98:23:98:23 | r | io.go:98:26:98:29 | &... |
 | io.go:98:26:98:29 | &... | io.go:97:7:97:9 | definition of buf |
[42/185 comp 12.9s eval 1s] FAILED(RESULT) /home/runner/work/codeql-go/codeql-go/ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.ql
```

I'll go ahead and regenerate the `.expected` file for `/home/runner/work/codeql-go/codeql-go/ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.ql`

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  private class FunctionModels extends TaintTracking::FunctionModel {
+    FunctionInput inp;
+    FunctionOutput outp;
+
+    FunctionModels() {
+      // signature: func NewReadWriter(r *Reader, w *Writer) *ReadWriter
+      hasQualifiedName("bufio", "NewReadWriter") and
+      (inp.isParameter(0) and outp.isResult())
+      or
+      inp.isResult() and outp.isParameter(1)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:18]:
```suggestion
      hasQualifiedName("bufio", "NewReadWriter") and
      (inp.isParameter(0) and outp.isResult()
      or
      inp.isResult() and outp.isParameter(1))
```

`or` binds less tighly than `and`

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      input = inp and output = outp
+    }
+  }
+
+  private class MethodModels extends TaintTracking::FunctionModel, Method {
+    FunctionInput inp;
+    FunctionOutput outp;
+
+    MethodModels() {
+      // Methods:
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:67]:
```suggestion
```

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      // signature: func (*Reader).Peek(n int) ([]byte, error)
+      this.(Method).hasQualifiedName("bufio", "Reader", "Peek") and
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:69]: No need to cast `this` to `Method` here and below; that's implied by `extends (...) Method`.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      (inp.isReceiver() and outp.isResult(0))
+      or
+      // signature: func (*Reader).ReadByte() (byte, error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:72]: I'd say that a single byte isn't enough to carry taint, so let's drop this unless we have strong evidence that this is a useful taint step.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      or
+      // signature: func (*Reader).ReadBytes(delim byte) ([]byte, error)
+      this.(Method).hasQualifiedName("bufio", "Reader", "ReadBytes") and
+      (inp.isReceiver() and outp.isResult(0))
+      or
+      // signature: func (*Reader).ReadLine() (line []byte, isPrefix bool, err error)
+      this.(Method).hasQualifiedName("bufio", "Reader", "ReadLine") and
+      (inp.isReceiver() and outp.isResult(0))
+      or
+      // signature: func (*Reader).ReadRune() (r rune, size int, err error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:84]: Ditto.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      or
+      // signature: func (*Writer).Reset(w io.Writer)
+      this.(Method).hasQualifiedName("bufio", "Writer", "Reset") and
+      (inp.isReceiver() and outp.isParameter(0))
+      or
+      // signature: func (*Writer).Write(p []byte) (nn int, err error)
+      this.(Method).hasQualifiedName("bufio", "Writer", "Write") and
+      (inp.isParameter(0) and outp.isReceiver())
+      or
+      // signature: func (*Writer).WriteByte(c byte) error
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:124]: Here as well.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      this.(Method).hasQualifiedName("bufio", "Writer", "WriteByte") and
+      (inp.isParameter(0) and outp.isReceiver())
+      or
+      // signature: func (*Writer).WriteRune(r rune) (size int, err error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bufio.qll:128]: And here.

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+      input = inp and output = outp
+    }
+  }
+
+  private class MethodModels extends TaintTracking::FunctionModel, Method {
+    FunctionInput inp;
+    FunctionOutput outp;
+
+    MethodModels() {
+      // Methods:
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:153]:
```suggestion
```

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+      // signature: func (*Buffer).Bytes() []byte
+      this.(Method).hasQualifiedName("bytes", "Buffer", "Bytes") and
+      (inp.isReceiver() and outp.isResult())
+      or
+      // signature: func (*Buffer).Next(n int) []byte
+      this.(Method).hasQualifiedName("bytes", "Buffer", "Next") and
+      (inp.isReceiver() and outp.isResult())
+      or
+      // signature: func (*Buffer).ReadByte() (byte, error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:162]: As above, I don't think this is very useful.

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+      or
+      // signature: func (*Buffer).ReadBytes(delim byte) (line []byte, err error)
+      this.(Method).hasQualifiedName("bytes", "Buffer", "ReadBytes") and
+      (inp.isReceiver() and outp.isResult(0))
+      or
+      // signature: func (*Buffer).ReadFrom(r io.Reader) (n int64, err error)
+      this.(Method).hasQualifiedName("bytes", "Buffer", "ReadFrom") and
+      (inp.isParameter(0) and outp.isReceiver())
+      or
+      // signature: func (*Buffer).ReadRune() (r rune, size int, err error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:174]: Ditto.

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+      or
+      // signature: func (*Buffer).String() string
+      this.(Method).hasQualifiedName("bytes", "Buffer", "String") and
+      (inp.isReceiver() and outp.isResult())
+      or
+      // signature: func (*Buffer).Write(p []byte) (n int, err error)
+      this.(Method).hasQualifiedName("bytes", "Buffer", "Write") and
+      (inp.isParameter(0) and outp.isReceiver())
+      or
+      // signature: func (*Buffer).WriteByte(c byte) error
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:190]: And again.

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+      this.(Method).hasQualifiedName("bytes", "Buffer", "WriteByte") and
+      (inp.isParameter(0) and outp.isReceiver())
+      or
+      // signature: func (*Buffer).WriteRune(r rune) (n int, err error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:194]: And another one.

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+      or
+      // signature: func (*Buffer).WriteTo(w io.Writer) (n int64, err error)
+      this.(Method).hasQualifiedName("bytes", "Buffer", "WriteTo") and
+      (inp.isReceiver() and outp.isParameter(0))
+      or
+      // signature: func (*Reader).ReadAt(b []byte, off int64) (n int, err error)
+      this.(Method).hasQualifiedName("bytes", "Reader", "ReadAt") and
+      (inp.isReceiver() and outp.isParameter(0))
+      or
+      // signature: func (*Reader).ReadByte() (byte, error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:210]: And one more.

File: ql/src/semmle/go/frameworks/stdlib/Bytes.qll [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+      this.(Method).hasQualifiedName("bytes", "Reader", "ReadByte") and
+      (inp.isReceiver() and outp.isResult(0))
+      or
+      // signature: func (*Reader).ReadRune() (ch rune, size int, err error)
```
Comment (max-schaefer) [inline ql/src/semmle/go/frameworks/stdlib/Bytes.qll:214]: And the last one.

File: ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.expected [hunk 15]
```diff
@@ -37,10 +37,12 @@
 | io.go:79:3:79:8 | reader | io.go:78:3:78:5 | definition of buf |
 | io.go:84:24:84:29 | reader | io.go:84:9:84:33 | call to LimitReader |
 | io.go:85:22:85:23 | lr | io.go:85:11:85:19 | selection of Stdout |
+| io.go:92:3:92:3 | definition of r | io.go:90:3:90:4 | definition of r2 |
```
Comment (max-schaefer) [inline ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.expected:40]: I'm not sure I understand this new result. Where did this edge come from? And why is it only for `r2`, bot not for `r1` and `r3`?

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll (hunk 4 above)
Comment (gagliardetto) [reply ql/src/semmle/go/frameworks/stdlib/Bufio.qll:72]: `ReadByte` and similar functions are (in most cases) used in a loop to gradually (a byte or rune at a time) read from the reader, and do some kind of filtering/processing and *assembling* of those bytes/runes.

The read bytes/runes in many (if not most) case do end up stitched together.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll (hunk 1 above)
Comment (gagliardetto) [reply ql/src/semmle/go/frameworks/stdlib/Bufio.qll]: Yep! Fixed.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll (hunk 3 above)
Comment (gagliardetto) [reply ql/src/semmle/go/frameworks/stdlib/Bufio.qll]: Removed all throughout.

File: ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.expected (hunk 15 above)
Comment (gagliardetto) [reply ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.expected:40]: I think these might be a side effect of https://github.com/github/codeql-go/pull/274#discussion_r465049692

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll (hunk 4 above)
Comment (max-schaefer) [reply ql/src/semmle/go/frameworks/stdlib/Bufio.qll:73]: > in many (if not most) case

That doesn't quite match my experience. Could you provide some evidence for this?

File: ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.expected (hunk 15 above)
Comment (gagliardetto) [reply ql/test/library-tests/semmle/go/frameworks/TaintSteps/TaintStep.expected]: Yep, adding the right brackets undid this.

File: ql/src/semmle/go/frameworks/stdlib/Bufio.qll (hunk 4 above)
Comment (gagliardetto) [reply ql/src/semmle/go/frameworks/stdlib/Bufio.qll]: You're right; the reassembly cases I had in mind do not provide a meaningful argument to include `ReadByte`, `ReadRune` and `WriteByte`, `WriteRune`.

**Comment by gagliardetto** [comment]
Great! Thanks.

**Pull request merged**